File: 3d/src/main/java/io/github/xerooup/frame3d/graphics/material/Material.java
package io.github.xerooup.frame3d.graphics.material;

import io.github.xerooup.frame3d.graphics.Color;

public class Material {
    private final MaterialType type;
    private Color color;
    private Texture texture;

    // Color material
    public Material(Color color) {
        this.type = MaterialType.COLOR;
        this.color = color;
    }

    // Texture material (UV mapped)
    public Material(Texture texture) {
        this.type = MaterialType.TEXTURE;
        this.texture = texture;
    }

    // Getters
    public MaterialType getType() {
        return type;
    }

    public Color getColor() {
        return color;
    }

    public Texture getTexture() {
        return texture;
    }

    // Factory methods
    public static Material color(Color color) {
        return new Material(color);
    }

    public static Material texture(String path) {
        return new Material(new Texture(path));
    }
}

File: 3d/src/main/java/io/github/xerooup/frame3d/graphics/material/Material.kt
package io.github.xerooup.frame3d.graphics.material

import io.github.xerooup.frame3d.graphics.Color

class Material {
    // Getters
    @JvmField
    val type: MaterialType
    var color: Color? = null
        private set
    var texture: Texture? = null
        private set

    // Color material
    constructor(color: Color?) {
        this.type = MaterialType.COLOR
        this.color = color
    }

    // Texture material (UV mapped)
    constructor(texture: Texture?) {
        this.type = MaterialType.TEXTURE
        this.texture = texture
    }

    companion object {
        // Factory methods
        fun color(color: Color?): Material {
            return Material(color)
        }

        fun texture(path: String?): Material {
            return Material(Texture(path))
        }
    }
}